### ISSUE TITLE
[BUG]: clang armv7-a uses an incorrect target

### ISSUE THREAD

**Issue by ghost** [compiler-explorer/compiler-explorer#3868]
### Describe the bug

The C++ compilers "clang armv7-a" are configured to use the target "arm-unknown-linux-gnueabi," yet from my understanding, this is an ARMv6 target, not an ARMv7 target. The proper target for ARMv7 should be "armv7-unknown-linux-gnueabi."

### Steps to reproduce

N/A?

### Expected behavior

The compiler emits ARMv7 codegen.

### Reproduction link

Not applicable

### Screenshots

Not applicable

**Comment by partouf** [comment]
Someone from ARM specifically added these to CE a couple years ago. (https://github.com/compiler-explorer/compiler-explorer/commit/7db8578bb5780ed0ee23d63e7277746940f2dc16)
Are you sure?

**Comment by partouf** [comment]
Ah you're right. @willlovett-arm changed these entries in this PR https://github.com/compiler-explorer/compiler-explorer/pull/2268/files , that's probably where it went wrong.

**Comment by partouf** [comment]
So perhaps we solved it in the wrong way and either should have;
* left it broken
* renamed them to armv6
* borrowed the correct v7 headers from the Android ndk or the official Arm compilers.

cc @mattgodbolt

**Comment by rearnsha** [comment]
I'm mostly a GCC for arm maintainer, but Will asked me if I could have a quick look at this.  I'm generally not convinced that trying to pick an architecture from the target string is a good idea.  The architecture selection is generally very complicated, with many extension options, so just saying 'armv7' is pretty meaningless.  Is that armv7, armv7 + div, armv7 + vfp, armv7 + simd, or some combination of those?  I presume clang has some configuration option to explicitly control the default target behaviour, though I've no idea how that works (gcc-8 and later have a '--with-arch=<specification>' to control this, though I guess this doesn't map well onto clang's multitarget model).

**Comment by partouf** [comment]
I agree that picking a target only is not the be-all-end-all, what you would usually do is at least also pick a CPU to target, then the compiler will figure out the most sensible defaults in terms of extensions.

But having Compiler Explorer pre-set the CPU is a bridge too far, we have to rely on the fact that people should know what kind of machine they actually want to target and preferable also know how to set that, as they're probably using the compiler in question themselves. Clang has a list of cpu's to choose from https://godbolt.org/z/64fTqb8eM

This issue is more about; does this arm7 what we offer at least vaguely in the direction of what you would assemble for a common cpu and platform.

I think when loading a basic example, there does not seem to be much difference; https://godbolt.org/z/qrjjM97PY
But when you use a more complicated example you do see a difference between targets 'arm' and 'armv7': https://godbolt.org/z/Krn4c5Ef7

I'm sure the arm example runs on a armv7 cpu (? I'm a x86 person, so I wouldn't know), but is that really the closest that we can get? armv7 would probably be a better target, and users can go more specific from there by setting the preferred CPU to target (and there are specific extension settings I'm sure).

But the next step to solve is that we're going to need header files that somewhat equal the architecture. So if we're targetting a 8 bits architecture with the compiler, the libc or c++ header files won't say that a pointer type is equal to int64_t or something.

Since Clang doesn't provide those header files, we have to improvise and fiddle around with some include paths. There's like system and kernel headers, libc headers, c++ headers. We know the Android NDK has specific header files for the Arm platform, because they have to work on all kinds of Android platforms.

But it's also likely that we could use one of the GCC Arm provided headers, I have not checked, but that might also not paint the whole picture.

Or maybe the best solution would be to download some basic packages for specific platforms of the linux version that we're using and at least get the linux, kernel, libc headers right.
